Implement device verification

TobiasFella opened the pull request:
Implements device verification using the to-device channel

The branch also contains the commits from #540 since it's easier for me to develop that way; only the last commit is relevant here though and it reviewing it only makes sense after #540 has been merged.

TobiasFella committed:
Add constructor for creating roomkeyevents

TobiasFella committed:
Implement sending encrypted messages

TobiasFella committed:
Save and load outgoing megolm session

TobiasFella committed:
Keep log of where we send keys and send keys to new devices and users

TobiasFella committed:
Implement sending encrypted files

TobiasFella committed:
Properly create encrypted edits

TobiasFella committed:
More work; Update olm pickle & timestamps in database; Remove TODOs

TobiasFella committed:
Update lib/events/encryptedfile.h

TobiasFella committed:
Update lib/room.cpp

TobiasFella committed:
Minor fix

TobiasFella committed:
Add database migration

TobiasFella committed:
Merge branch 'dev' into sendmessages

TobiasFella committed:
Fixes

TobiasFella committed:
Implement key verification

CarlSchwan commented on res.qrc:
```diff
@@ -0,0 +1,5 @@
+<!DOCTYPE RCC><RCC version="1.0">
+<qresource prefix="/">
```
maybe add a prefix?

CarlSchwan commented on lib/keyverificationsession.h:
```diff
@@ -0,0 +1,140 @@
+
+public:
+    enum State {
+        INCOMING, // There is a request for verification incoming
+        WAITINGFORREADY, // We sent a request for verification and are waiting for ready
+        READY, // Either party sent a ready as a response to a request; The user selects a method
+        WAITINGFORACCEPT, // We sent a start and are waiting for an accept
+        ACCEPTED, // The other party sent an accept and is waiting for a key
+        WAITINGFORKEY, // We're waiting for a key
+        WAITINGFORVERIFICATION, // We're waiting for the *user* to verify the emojis
+        WAITINGFORMAC, // We're waiting for the mac
+        CANCELED, // The session has been canceled
+        DONE, // The verification is done
```
```suggestion
        INCOMING, ///< There is a incoming request for verification
        WAITINGFORREADY, ///< We sent a request for verification and are waiting for ready
        READY, ///< Either party sent a ready as a response to a request; The user selects a method
        WAITINGFORACCEPT, ///< We sent a start and are waiting for an accept
        ACCEPTED, ///< The other party sent an accept message and is waiting for a key
        WAITINGFORKEY, ///< We're waiting for a key
        WAITINGFORVERIFICATION, ///< We're waiting for the *user* to verify the emojis
        WAITINGFORMAC, ///< We're waiting for the mac
        CANCELED, ///< The session has been canceled
        DONE, ///< The verification is done
```

CarlSchwan commented on lib/events/keyverificationevent.cpp:
```diff
@@ -162,3 +162,33 @@ QHash<QString, QString> KeyVerificationMacEvent::mac() const
+    return contentPart<QString>("transaction_id"_ls);
+}
+
+
```
```suggestion
```

CarlSchwan commented on lib/events/encryptedevent.h:
```diff
@@ -56,6 +56,10 @@ class QUOTIENT_API EncryptedEvent : public RoomEvent {
+
+    void setRelation(const QJsonObject& relation);
+
+    bool isVerified();
```
not implemented?

CarlSchwan commented on lib/connection.cpp:
```diff
@@ -2235,3 +2332,58 @@ bool Connection::isKnownCurveKey(const QString& user, const QString& curveKey)
+void Connection::startKeyVerificationSession(const QString& deviceId)
+{
+    auto session = new KeyVerificationSession(userId(), deviceId, this, this);
+    Q_EMIT newKeyVerificationSession(session);
```
```suggestion
    emit newKeyVerificationSession(session);
```

CarlSchwan commented on lib/connection.cpp:
```diff
@@ -2235,3 +2332,58 @@ bool Connection::isKnownCurveKey(const QString& user, const QString& curveKey)
+    if (!query.next()) {
+        return false;
+    }
+    auto curveKey = query.value("senderKey"_ls).toString();
```
```suggestion
    const auto curveKey = query.value("senderKey"_ls).toString();
```

CarlSchwan commented on lib/connection.cpp:
```diff
@@ -2235,3 +2332,58 @@ bool Connection::isKnownCurveKey(const QString& user, const QString& curveKey)
+        payloadJson["keys"] = senderObject;
+
+        const auto& u = user(userId);
+        auto cipherText = olmEncryptMessage(u, deviceId, QJsonDocument(payloadJson).toJson(QJsonDocument::Compact));
```
```suggestion
        const auto cipherText = olmEncryptMessage(u, deviceId, QJsonDocument(payloadJson).toJson(QJsonDocument::Compact));
```

KitsuneRal commented on lib/connection.cpp:
```diff
@@ -2297,3 +2332,58 @@ bool Connection::isKnownCurveKey(const QString& user, const QString& curveKey)
+        payloadJson["keys"] = senderObject;
+
+        const auto& u = user(userId);
+        auto cipherText = olmEncryptMessage(u, deviceId, QJsonDocument(payloadJson).toJson(QJsonDocument::Compact));
```
At least a part of that looks worth factoring out into some `Connection::encryptEvent` method - we already have similar code involving `RoomKeyEvent`.

KitsuneRal commented on lib/connection.cpp:
```diff
@@ -974,8 +976,23 @@ void Connection::Private::consumeToDeviceEvents(Events&& toDeviceEvents)
             pendingEncryptedEvents.push_back(std::make_unique<EncryptedEvent>(event.fullJson()));
-        }, [](const Event& e){
-            // Unhandled
+        }, [this](const KeyVerificationRequestEvent& event) {
+            auto session = new KeyVerificationSession(q->userId(), event, q, false, q);
+            emit q->newKeyVerificationSession(session);
+        }, [this](const KeyVerificationReadyEvent& event) {
+            emit q->incomingKeyVerificationReady(event);
+        }, [this](const KeyVerificationStartEvent& event) {
+            emit q->incomingKeyVerificationStart(event);
+        }, [this](const KeyVerificationAcceptEvent& event) {
+            emit q->incomingKeyVerificationAccept(event);
+        }, [this](const KeyVerificationKeyEvent& event) {
+            emit q->incomingKeyVerificationKey(event);
+        }, [this](const KeyVerificationMacEvent& event) {
+            emit q->incomingKeyVerificationMac(event);
+        }, [this](const KeyVerificationDoneEvent& event) {
+            emit q->incomingKeyVerificationDone(event);
+        }, [this](const KeyVerificationCancelEvent& event) {
+            emit q->incomingKeyVerificationCancel(event);
```
I wonder if it makes sense to replace this switchboard with one signal to announce changes in the verification session state machine. By the way, you can only pass event (references to events, that is) through direct signal/slot connections - probably it's better to update the the verification session and pass a pointer to it in the signal instead?

KitsuneRal commented on lib/connection.cpp:
```diff
@@ -998,9 +1015,25 @@ void Connection::Private::handleEncryptedToDeviceEvent(const EncryptedEvent& eve
-        },
-        [](const Event& evt) {
-            qCDebug(E2EE) << "Skipping encrypted to_device event, type"
+        }, [this](const KeyVerificationRequestEvent& event) {
+            auto session = new KeyVerificationSession(q->userId(), event, q, true, q);
+            emit q->newKeyVerificationSession(session);
+        }, [this](const KeyVerificationReadyEvent& event) {
+            emit q->incomingKeyVerificationReady(event);
+        }, [this](const KeyVerificationStartEvent& event) {
+            emit q->incomingKeyVerificationStart(event);
+        }, [this](const KeyVerificationAcceptEvent& event) {
+            emit q->incomingKeyVerificationAccept(event);
+        }, [this](const KeyVerificationKeyEvent& event) {
+            emit q->incomingKeyVerificationKey(event);
+        }, [this](const KeyVerificationMacEvent& event) {
+            emit q->incomingKeyVerificationMac(event);
+        }, [this](const KeyVerificationDoneEvent& event) {
+            emit q->incomingKeyVerificationDone(event);
+        }, [this](const KeyVerificationCancelEvent& event) {
+            emit q->incomingKeyVerificationCancel(event);
```
Same as above.

KitsuneRal reviewed (changes requested):
Apologies for a terribly long wait - can you look at reducing the repetitive code, as per below?

KitsuneRal commented on lib/keyverificationsession.cpp:
```diff
@@ -0,0 +1,522 @@
+{
+    static QJsonArray data;
+    if (data.isEmpty()) {
+        QFile dataFile(":/sas-emoji.json");
```
I think it's better to keep this file out on the file system rather than swallow it in a Qt resource, so that it could be updated independently from the client (and possibly even automatically downloaded by the client whenever a new version is found).

KitsuneRal committed:
Connection::Private::assembleEncryptedContent()

What was partially factored out before into encryptSessionKeyEvent()
is now the complete algorithm converting any event json into encrypted
content.

KitsuneRal committed:
Merge branch 'dev' into device-verification

# Conflicts:
#	autotests/testfilecrypto.cpp
#	lib/connection.cpp
#	lib/connection.h
#	lib/database.cpp
#	lib/database.h
#	lib/e2ee/qolmoutboundsession.cpp
#	lib/e2ee/qolmoutboundsession.h
#	lib/eventitem.h
#	lib/events/encryptedevent.cpp
#	lib/events/encryptedevent.h
#	lib/events/encryptedfile.cpp
#	lib/events/encryptedfile.h
#	lib/events/keyverificationevent.cpp
#	lib/events/keyverificationevent.h
#	lib/events/roomkeyevent.h
#	lib/room.cpp
#	lib/room.h

KitsuneRal commented:
After getting lost in trying to work out the next-gen event type infrastructure, I think I'll just merge this PR with some cleanup and minor refactoring - assuming that it can be tested with some unpublished code (not mine) :) Ideally we should have a separate test in Quotest that would go through at least some verification states; but in order not to hold this back behind myself any longer, this can be postponed to a separate PR.

KitsuneRal committed:
KeyVerificationSession: cleanup

- Use std::chrono for the timeout (it's more readable and
  less ambiguous) and make it a local variable
- Only pass a Connection object once to constructors
- Ensure buildability even without E2EE (key verification is disabled
  in that case)
- Reorder #includes
- Other cleanup following clang-tidy warnings

KitsuneRal committed:
More code reorganisation

- Common switchOnType() piece for key verification events is factored
  out into processIfVerificationEvent()
- Bare event JSON removed from KeyVerificationSession into constructors
  of respective events
- Connection::sendToDevice() uses assembleEncryptedContent() introduced
  in the previous commit
- commonSupportedMethods() moved out to .cpp; error/string converters
  made static

KitsuneRal committed:
Refactor the code handling emoji

- Use a dedicated structure, EmojiEntry, instead of QVariantMap
  (it's Q_GADGET, should be readable from QML just fine)
- While we're at it, QVector is better than QList in Qt 5.15
- Remove language from the session state - it's used in a single method
- Modernise handleKey() code
